SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,625 @@
+                    return true
+                }
+
+                if let faiCallee = fai.referencedFunction {
+                    if faiCallee.isAvailableExternally && 
+                    !handledFuncs.contains(faiCallee) && 
+                    handledFuncs.count < recursionBudget {
+                        handledFuncs.insert(faiCallee)
+                        if inner(faiCallee, fai.calleeArgumentIndex(of: use)!, &handledFuncs) {
+                            return true
+                        }
+                    }
+                }
+            }
+        }
+
+        return false
+    }
+
+    // Limit the number of recursive calls to not go into 
+    // exponential behavior in corner cases.
+    let recursionBudget = 8
+    var handledFuncs: Set<Function> = []
+    return inner(callee, index, &handledFuncs)
+}
+
+/// Marks any intermediate reabstracted closures corresponding to given closure
+/// passed in at a callsite as used. We do not want to look at reabstraction closures 
+/// twice as they are covered as part of the final closure that we're passing in at the  
+/// callsite.
+func markReabstractionPartialApplyAsUsed(firstClosure: Value, currentClosure: Value, usedReabstractionClosures: inout InstructionSet) {
+    if currentClosure == firstClosure {
+        return
+    }
+
+    if let pai = currentClosure as? PartialApplyInst {
+        usedReabstractionClosures.insert(pai)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: pai.callee, usedReabstractionClosures: &usedReabstractionClosures)
+    }
+
+    if let cvt = currentClosure as? ConvertFunctionInst {
+        usedReabstractionClosures.insert(cvt)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: cvt.fromFunction, usedReabstractionClosures: &usedReabstractionClosures)
+    }
+
+    if let cvt = currentClosure as? ConvertEscapeToNoEscapeInst {
+        usedReabstractionClosures.insert(cvt)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: cvt.fromFunction, usedReabstractionClosures: &usedReabstractionClosures)
+    }
+
+    if let mdi = currentClosure as? MarkDependenceInst {
+        usedReabstractionClosures.insert(mdi)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: mdi.value, usedReabstractionClosures: &usedReabstractionClosures)
+    }   
+
+    fatalError("Unexpected instruction used during closure reabstraction: \(currentClosure)")
+}
+
+// ===================== Types ===================== //
+
+/// Represents all the information required to
+/// represent a closure in isolation, i.e., outside of
+/// a callsite context where the closure may be getting
+/// passed as an argument.
+///
+/// Composed with other information inside a `ClosureArgDescriptor`
+/// to represent a closure as an argument at a callsite.
+class ClosureInfo {
```

Comment:
Is this a class because it can be contained in multiple `ClosureArgDescriptor`s?

Reply:
Yes. And because of the way `lifetimeFrontier` is set. We set it after multiple closure arg descriptors have already been given a closure info instance.

Are there some drawbacks to making this a class?

Reply:
I've changed it to a struct in the process of separating things out from the `gatherCallSites` function. I'm still curious to know if there would have been some drawbacks to making this a class?

Reply:
> Are there some drawbacks to making this a class?

Classes have a performance overhead: memory allocation and reference counting

But if it's the right choice then it should be a class. If it turns out to have a negative performance impact, we can optimize it later.

Reply:
Also, you already have a lot of implicit memory allocations in the pass. Every Array and Dictionary will call malloc.
That's probably fine. If not, we can deal with it later.

Reply:
I initially saw some issues when I was trying to store handles to `InstructionRange` and I may have taken the "don't store any of the custom data structures" too far. 

Do you have any guidance on that?